Data/HashMap/Internal.hs

```diff
@@ -196,8 +205,10 @@ import Data.Coerce (coerce)
 ------------------------------------------------------------------------
 
 -- | Convenience function.  Compute a hash value for the given value.
-hash :: H.Hashable a => a -> Hash
-hash = fromIntegral . H.hash
+hash ::forall a (salt :: Nat) . H.Hashable a => KnownNat salt => Proxy salt -> a -> Hash
+hash proxy = fromIntegral . H.hashWithSalt (fromIntegral salt)
+   where
+     salt = natVal proxy -- TODO ensure is unboxed, ensure is a Word
```

Comment:
I very much doubt that it's possible to unbox it with current GHC, but I'd love to be proven wrong.

Question: would we get better code/specialization/performance with an auxiliary class? I'm not at all sure, but it seems vaguely plausible.

```haskell
wordVal :: forall n. KnownWord n => Proxy# n -> Word
wordVal _ = unTagged (wordVal' :: Tagged n Word)

class KnownWord (n :: Nat) where
  wordVal' :: Tagged n Word
instance KnownNat n => KnownWord n where
#if MIN_VERSION_base(4,8,0)
  wordVal = Tagged $ fromIntegral (natVal' (proxy# :: Proxy# n))
#else
  wordVal _ = Tagged $ fromIntegral (natVal (Proxy :: Proxy n))
#endif
```

Reply:
I'm a bit confused about this, how do I know if the resulting code boxes it or not?
My understanding of boxing is in C:

```
int* salt = malloc(sizeof(int)); // reserves memory somehwere, pointer gets pushed on stack
(*salt) = 50; // put 50 into memory somehwere using pointer
```
rather then

```
int salt = 50; // put on stack
```

How do I make ghc tell me which of those two options it chooses?